petpal/seekers/views.py

```diff
@@ -13,24 +21,197 @@
+@permission_classes([IsAuthenticated])
 def seeker_detail_view(request, account_id):
+    # Check that the seeker with account_id exists
+    try:
+        seeker = Seeker.objects.get(id=account_id)
+    except Seeker.DoesNotExist:
+        return Response({'error': 'User not found'}, status=status.HTTP_404_NOT_FOUND)
+
     if request.method == 'GET':
-        return Response({'msg': 'seeker detail'}, status=status.HTTP_200_OK)
+        # Check that the request is from the user themselves or a shelter.
+        if request.user == seeker or request.user.role == PetPalUser.Role.SHELTER:
+            # If the request is from a shelter, check it has an application with the seeker
+            if request.user.role == PetPalUser.Role.SHELTER:
+                try:
+                    Application.objects.get(seeker=account_id, shelter=request.user)
+                except Application.DoesNotExist:
+                    return Response({'error': 'You are not allowed to see this profile'},
+                                    status=status.HTTP_403_FORBIDDEN)
+
+            # Retrieves the following data: 'email', 'address', 'city', 'province', 'postal_code', 'phone'.
+            serializer = SeekerSerializer(seeker, many=False)
+
+            # Retrieve the favorite pets name
+            try:
+                # fav_pets: a list of dictionaries
+                fav_pets = PetListing.objects.filter(favorited_by=seeker)
+            except PetListing.DoesNotExist:
+                fav_pets = None
+
+            # Case if there are no favorite pets
+            if fav_pets is None:
+                return Response({
+                    'msg': 'Seeker details',
+                    'user_data': serializer.data,
+                    'fav_pets': 'No favorites yet!'}, status=status.HTTP_200_OK)
+            else:
+                # Serialize the favorite pets
+                fav_pet_list = []
+                for pet in fav_pets.values():
+                    # pet_serialized = serialize_fav_pets(pet)
+                    pet_serialized = FavPetSerializer(pet, many=False)
+                    fav_pet_list.append(pet_serialized.data)
+
+                return Response({
+                    'msg': 'Seeker details',
+                    'user_data': serializer.data,
+                    'fav_pets': fav_pet_list}, status=status.HTTP_200_OK)
+        else:
+            return Response({'error': 'You are not allowed to see this profile'}, status=status.HTTP_403_FORBIDDEN)
 
     if request.method == 'PUT':
-        return Response({'msg': 'edit seeker'}, status=status.HTTP_200_OK)
-    
-    if request.method == 'DELETE':
-        return Response({'msg': 'delete seeker'}, status=status.HTTP_200_OK)
+        # Check that the request is from the user themselves
+        if request.user == seeker:
+            serializer = SeekerSerializer(seeker, data=request.data, partial=True)
+            if serializer.is_valid():
+                serializer.save()
```

Comment:
Can you also update the notion for the payload of this endpoint?

Reply:
Sure!